File: plugins/reins/src/stream.ts
/**
 * Stream proxy: formats claude-agent-sdk messages for UI display.
 *
 * SDK message types:
 *   SDKAssistantMessage  — { type:'assistant', message: BetaMessage }
 *   SDKUserMessage       — { type:'user', message: MessageParam }
 *   SDKResultMessage     — { type:'result', subtype:'success'|'error', ... }
 *   SDKSystemMessage     — { type:'system', subtype:'init', ... }
 *   SDKRateLimitEvent    — { type:'rate_limit_event', ... }
 */

export function formatSdkMessage(msg: any, _agentName: string): string | null {
  if (msg.type === "assistant") {
    const parts: string[] = [];
    // msg.message is a BetaMessage; content blocks are in msg.message.content
    for (const block of msg.message?.content ?? []) {
      if (block.type === "thinking") {
        const text = (block.thinking ?? "").slice(0, 300);
        const suffix = (block.thinking ?? "").length > 300 ? "..." : "";
        parts.push(`\u{1F4AD} *${text}${suffix}*`);
      } else if (block.type === "text") {
        parts.push(block.text);
      } else if (block.type === "tool_use") {
        let inputStr = JSON.stringify(block.input ?? {});
        if (inputStr.length > 150) inputStr = inputStr.slice(0, 147) + "...";
        parts.push(`\u{1F527} **${block.name}**\n\`\`\`json\n${inputStr}\n\`\`\``);
      }
    }
    return parts.length > 0 ? parts.join("\n\n") : null;
  }

  // UserMessage — tool results
  if (msg.type === "user") {
    const content = msg.message?.content;
    if (Array.isArray(content)) {
      for (const block of content) {
        if (block.type === "tool_result") {
          const text = String(block.content ?? "").slice(0, 500);
          if (block.is_error) {
            return `\u274C **Error:**\n\`\`\`\n${text}\n\`\`\``;
          }
          return `\`\`\`\n${text}\n\`\`\``;
        }
      }
    }
    return null;
  }

  // ResultMessage
  if (msg.type === "result") {
    const duration = msg.duration_ms ? ` (${(msg.duration_ms / 1000).toFixed(1)}s)` : "";
    const turns = msg.num_turns ? `, ${msg.num_turns} turns` : "";
    if (msg.is_error) {
      return `\u26A0\uFE0F **Ended:** \`${msg.subtype}\`${duration}${turns}`;
    }
    return `\u2705 **Done**${duration}${turns}`;
  }

  // SystemMessage
  if (msg.type === "system") {
    if (msg.subtype === "init") return "\uD83D\uDCC2 Session started";
    return null;
  }

  // Ignore rate limits, status, hooks, etc.
  return null;
}

File: scripts/stream.py
"""Stream proxy: formats ClaudeSDKClient messages for Gradio display."""

import json
from claude_agent_sdk import (
    AssistantMessage, ResultMessage, SystemMessage,
    UserMessage, TextBlock, ToolUseBlock, ToolResultBlock,
    ThinkingBlock, RateLimitEvent,
)


def format_sdk_message(msg, agent_name: str) -> str | None:
    """Format a single SDK stream message for Gradio chat display.
    Returns formatted string or None (skip this message).
    """
    if isinstance(msg, AssistantMessage):
        parts = []
        for block in msg.content:
            if isinstance(block, ThinkingBlock):
                text = block.thinking[:300]
                suffix = "..." if len(block.thinking) > 300 else ""
                parts.append(f"💭 *{text}{suffix}*")
            elif isinstance(block, TextBlock):
                parts.append(block.text)
            elif isinstance(block, ToolUseBlock):
                input_str = json.dumps(block.input, ensure_ascii=False)
                if len(input_str) > 150:
                    input_str = input_str[:147] + "..."
                parts.append(f"🔧 **{block.name}**\n```json\n{input_str}\n```")
        return "\n\n".join(parts) if parts else None

    if isinstance(msg, UserMessage):
        for block in msg.content:
            if isinstance(block, ToolResultBlock):
                content = str(block.content)[:500]
                if block.is_error:
                    return f"❌ **Error:**\n```\n{content}\n```"
                return f"```\n{content}\n```"
        return None

    if isinstance(msg, ResultMessage):
        duration = f" ({msg.duration_ms / 1000:.1f}s)" if msg.duration_ms else ""
        turns = f", {msg.num_turns} turns" if msg.num_turns else ""
        if msg.is_error:
            return f"⚠️ **Ended:** `{msg.subtype}`{duration}{turns}"
        return f"✅ **Done**{duration}{turns}"

    if isinstance(msg, SystemMessage):
        if msg.subtype == "init":
            return "📂 Session started"
        return None

    if isinstance(msg, RateLimitEvent):
        return None

    return None
